feat: Add upcoming bookings feature to GET /api/bookings endpoint

ImBIOS opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #15983
- Added support for filtering upcoming bookings by adding the `upcoming` query parameter to the GET /api/bookings endpoint.
- When `upcoming` is set to "true", only upcoming bookings are returned.
- Regular users can now retrieve only their upcoming bookings by setting `upcoming` to "true".
- System-wide admins and organization admins can also retrieve only upcoming bookings by setting `upcoming` to "true".

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Are there environment variables that should be set? nope
- What are the minimal test data to have? ???
- What is expected (happy path) to have (input and output)? ???
- Any other important info that could help to test that PR :thinking:

ImBIOS committed:
feat: Add upcoming bookings feature to GET /api/bookings endpoint

- Added support for filtering upcoming bookings by adding the `upcoming` query parameter to the GET /api/bookings endpoint.
- When `upcoming` is set to "true", only upcoming bookings are returned.
- Regular users can now retrieve only their upcoming bookings by setting `upcoming` to "true".
- System-wide admins and organization admins can also retrieve only upcoming bookings by setting `upcoming` to "true".

ImBIOS commented:
https://cal.com/docs/enterprise-features/api/api-reference/bookings#find-all-bookings

Should I update something on this page? where is this code located?

anikdhabal commented:
Thanks for your work @ImBIOS , can you share a loom video of the result?

ImBIOS commented:
@anikdhabal what should I include in the recording with Loom? Do you know how can I run integration-test I just wrote in this PR?

anikdhabal commented:
> @anikdhabal what should I include in the recording with Loom? Do you know how can I run integration-test I just wrote in this PR?

Did you test that locally and ensure that it works? Also, you can run this command - `yarn test -- --integrationTestsOnly` for integration tests.

supalarry commented on apps/api/v1/pages/api/bookings/_get.ts:
```diff
@@ -130,25 +137,30 @@ type GetAdminArgsType = {
   userIds: number[] = [],
-  userEmails: string[] = []
+  userEmails: string[] = [],
+  upcomingOnly: boolean
```
Like I described in other comment - we can pass 1 prop `status: Status` and then  if status === "upcoming" we add the where clause.

supalarry commented on apps/api/v1/pages/api/bookings/_get.ts:
```diff
@@ -177,9 +189,15 @@ function buildWhereClause(
-  return {
-    ...whereClause,
-  };
+  // Add upcoming filter if requested
```
There is no need to add `// Add upcoming filter if requested` comment when code explains it already - when reading `if (upcomingOnly) {` it is already clear what will happen : )

supalarry commented on apps/api/v1/test/lib/bookings/_get.integration-test.ts:
```diff
@@ -105,4 +105,89 @@ describe("GET /api/bookings", async () => {
     expect(groupedUsers.size).toBeGreaterThanOrEqual(2);
   });
+
+  // Tests for upcoming bookings feature - START
```
Instead of using `// Tests for upcoming bookings feature - START` and // Tests for upcoming bookings feature - END  it is better to group all of these `it` tests into a `describe` block:

```
describe("test upcoming bookings filter", () => {
    it("Returns only upcoming bookings when upcoming=true for regular user", async () => {
        ...
    });
    ...
});
```

supalarry commented on apps/api/v1/lib/validations/booking.ts:
```diff
@@ -23,6 +23,7 @@ export const schemaBookingGetParams = z.object({
   dateTo: iso8601.optional(),
   order: z.enum(["asc", "desc"]).default("asc"),
   sortBy: z.enum(["createdAt", "updatedAt"]).optional(),
+  upcoming: z.enum(["true", "false"]).optional(),
```
I think it's better to have type Status = "upcoming" to which we later on can add other statuses like type Status = "upcoming" | "past" and then use it as status: Status.

Rationale: it is very likely we will want to filter also by past, cancelled, recurring etc. so we would have to add many properties like pastOnly, and instead we can replace that logic with 1 property status: Status.

supalarry replied:
Ideally this would be an array `status: Status[]` that would allow passing multiple statuses, but for now `status: Status` should be good enough because in the future we will migrate to another api version (this is version 1 and we are working on version 2)

supalarry reviewed (changes requested):
Great PR Imamuzzaki! Small things need to be adjusted and then it will be perfect.

Let me know what you think:

anikdhabal reviewed (changes requested):
Integration test is failing

ImBIOS committed:
Merge branch 'main' into feat/add-upcoming-booking-v1

ImBIOS committed:
feat: Add status filter to booking API endpoint

This commit adds a new `status` filter to the booking API endpoint. The `status` filter allows filtering bookings by their status, overriding the `dateFrom` and `dateTo` filters. The only valid value for the `status` filter is "upcoming".

anikdhabal commented on apps/api/v1/pages/api/bookings/_get.ts:
```diff
@@ -63,6 +63,25 @@ import { schemaQuerySingleOrMultipleUserIds } from "~/lib/validations/shared/que
+ *         required: false
+ *         schema:
+ *          type: string
+ *          enum: [upcoming]
+ *          description: Filter bookings by status, it wil ocerwrite dateFrom and dateTo filters
```
Fix the typo in description

anikdhabal committed:
Update _get.ts

anikdhabal reviewed (approved):
LGTM !! Need approval form @supalarry

ImBIOS commented:
If v2 is WIP, maybe someone needs to add this feature to that v2 TODO list.
